src/v/cluster/rm_stm.cc

```diff
@@ -1307,20 +1235,43 @@ ss::future<result<kafka_result>> rm_stm::transactional_replicate(
           _ctx_log.warn,
           "application of the replicated tx batch has timed out pid:{}",
           bid.pid);
-        if (_raft->is_leader() && _raft->term() == synced_term) {
-            co_await _raft->step_down("replicate_tx wait error");
-        }
+        req_ptr->set_value(errc::timeout);
         co_return tx_errc::timeout;
     }
-
-    auto old_offset = r.value().last_offset;
-    auto new_offset = from_log_offset(old_offset);
-
-    set_seq(bid, new_offset);
-
     _mem_state.estimated.erase(bid.pid);
+    auto result = kafka_result{
+      .last_offset = from_log_offset(r.value().last_offset)};
+    req_ptr->set_value(result);
+    co_return result;
+}
+
+ss::future<result<kafka_result>> rm_stm::transactional_replicate(
+  model::batch_identity bid, model::record_batch_reader rdr) {
+    if (!check_tx_permitted()) {
+        co_return errc::generic_tx_error;
+    }
+    auto producer = maybe_create_producer(bid.pid);
+    auto result = co_await producer->run_func([&](ssx::semaphore_units units) {
+        return do_transactional_replicate(
+          producer, bid, std::move(rdr), std::move(units));
+    });
 
-    co_return kafka_result{.last_offset = new_offset};
+    if (!result) {
+        if (result.error() == errc::sequence_out_of_order) {
+            auto barrier = co_await _raft->linearizable_barrier();
+            if (!barrier) {
+                co_return errc::not_leader;
+            }
+        }
+        if (
+          _raft->is_leader()
+          && (result.error() == errc::replication_error || result.error() == errc::timeout)) {
```

Comment:
wrong condition

Reply:
fixed.

Reply:
it's super dangerous, by default any replication error must result in stepping down in order to preserve correctness. here we step down only on timeout.

Why do you think it's the only error code with unclear outcome of replication? (Assuming it's true) can we guarantee that it remains the only error code with such behavior in the future?

It's safe to assume for an error with clear outcome that it's uncertain but the opposite isn't true so we should assume bad case and filter good scenarios than the opposite.

Reply:
`do_idempotent_replicate` has this return branch `co_return result.error();` so not all replication errors are ` errc::replication_error`

Reply:
is it really super dangerous, this step_down is executed outside of the lock and doesn't guarantee that all requests which were enqued after the current that failed are failed as there are multiple scheduling points in between, also in the previous implementation the comment was suggesting that the stepdown is: `just in case stepping down to minimize the risk`